docs/CppInteroperability/Roadmap.md

```diff
@@ -0,0 +1,141 @@
+# Swift and C++ Interoperability Roadmap
+
+* Introduction
+    * C++, as a language
+    * C and Objective-C interoperability
+* Specifics and trade-offs
+* Goals
+* The approach 
+    * Foreign reference types
+    * Owned types and value lifetimes
+    * Iterators and ranges (by Egor)
+    * Generic APIs https://forums.swift.org/t/bridging-c-templates-with-interop/55003
+* Evolution and process
+    * Example: virtual/overridden APIs
+* How interop fits into the ecosystem more generally (lldb, ide, spm)
+    * *The* Standard Library overlay (by Egor)
+
+## Introduction
+
+**This roadmap is** a sketch, rather than a final design for the “forward” (using C++ APIs from Swift) half of C++ and Swift interoperability. This roadmap outlines some high-level topics related to C++ interoperability, overarching goals that drive the project’s design decisions, a process for evolving C++ interoperability over time, and a collection of specific problems and their potential solutions. 
+
+“Reverse” interoperability (using Swift APIs from C++) is another extremely important part of the interoperability story, however, reverse interoperability has largely different goals and constraints, which necessarily mean a different design and roadmap. Alex Lorenz will be putting together a roadmap for reverse interoperability which outlines these goals and will shed light onto the differences between the two halves of the projects. In the future this roadmap will need to be updated to explain how these two halves fit together and present the “bi-directional” interoperability story that is C++ and Swift interoperability as a whole. 
```

Comment:
> different goals and constraints

Maybe expand on these constrains and goals: going from some specific Swift API to a broad language with less strong idioms. This actually gives a lot for flexibility for the interop model.

Reply:
I disagree that these things are very different.  As soon as you can pass function pointers or closures across the language boundary, you effectively have the problems of bidirectional interop.  Not to be alarmist, but tackling the two directions independently with the assumption that they are fundamentally different seems like a disaster waiting to happen.

Reply:
This also speaks to the implict/explicit conversion discussion we had today. If I pass (in a swift file) a swift string to a C++ API that expects a std::string, that C++ implementation is going to go off and make assumptions about how it can edit and access the data. So we have to think about how C++ would use Swift data types, not just how Swift would use C++ data types.